packages/app/src/ui/pages/PresentVerifiableCredential/usePresentVerifiableCredential.ts

```diff
@@ -0,0 +1,176 @@
+import { IVerifiablePresentation } from "@cryptkeeperzk/types";
+import { useCallback, useEffect, useState } from "react";
+
+import {
+  generateVerifiablePresentationFromVerifiableCredentials,
+  serializeVerifiablePresentation,
+} from "@src/background/services/credentials/utils";
+import { ICryptkeeperVerifiableCredential } from "@src/types";
+import { closePopup } from "@src/ui/ducks/app";
+import { useAppDispatch } from "@src/ui/ducks/hooks";
+import {
+  fetchVerifiableCredentials,
+  rejectVerifiablePresentationRequest,
+  generateVerifiablePresentation,
+} from "@src/ui/ducks/verifiableCredentials";
+import { useCryptkeeperVerifiableCredentials } from "@src/ui/hooks/verifiableCredentials";
+import { useEthWallet } from "@src/ui/hooks/wallet";
+
+const ETHEREUM_SIGNATURE_SPECIFICATION_TYPE = "EthereumEip712Signature2021";
+const VERIFIABLE_CREDENTIAL_PROOF_PURPOSE = "assertionMethod";
+
+export interface IUsePresentVerifiableCredentialData {
+  isWalletConnected: boolean;
+  isWalletInstalled: boolean;
+  verifiablePresentationRequest?: string;
+  cryptkeeperVerifiableCredentials: ICryptkeeperVerifiableCredential[];
+  selectedVerifiableCredentialHashes: string[];
+  verifiablePresentation?: IVerifiablePresentation;
+  error?: string;
+  onCloseModal: () => void;
+  onRejectRequest: () => void;
+  onToggleSelection: (hash: string) => void;
+  onConfirmSelection: () => void;
+  onReturnToSelection: () => void;
+  onConnectWallet: () => Promise<void>;
+  onSubmitVerifiablePresentation: (needsSignature: boolean) => Promise<void>;
+}
+
+export const usePresentVerifiableCredential = (): IUsePresentVerifiableCredentialData => {
+  const [verifiablePresentationRequest, setVerifiablePresentationRequest] = useState<string>();
+  const cryptkeeperVerifiableCredentials = useCryptkeeperVerifiableCredentials();
+  const [selectedVerifiableCredentialHashes, setSelectedVerifiableCredentialHashes] = useState<string[]>([]);
+  const [verifiablePresentation, setVerifiablePresentation] = useState<IVerifiablePresentation>();
+  const [error, setError] = useState<string>();
+
+  const ethWallet = useEthWallet();
+  const dispatch = useAppDispatch();
+
+  useEffect(() => {
+    function getPresentationRequest() {
+      const { searchParams } = new URL(window.location.href.replace("#", ""));
+      const request = searchParams.get("request");
+
+      if (!request) {
+        return;
+      }
+
+      dispatch(fetchVerifiableCredentials());
+
+      setVerifiablePresentationRequest(request);
+    }
+    getPresentationRequest();
+  }, [setVerifiablePresentationRequest, fetchVerifiableCredentials, dispatch]);
```

Comment:
This effect won't work properly if url is changed. Also, you don't need to wrap it with function and call it immediately.

Reply:
I fixed it so that there is no function wrap anymore. I'm not sure the Url ever will change in practice. This is the same as for `AddVerifiableCredential` - these are not meant to be screens that are ever shown in their own tab, as we discussed in the last PR, they are simply popups and will be separate from your feature which opens pages in a new tab.

Reply:
If user opens these popups in a new tab manually, it has style issues. In addition, there is a task to support event propagation to multiple tabs which allows user to use the extension in a separate tab or window without closing it. There is no guarantee that it won't be shown in separate tab so I'd fix style issues for this case.

Reply:
Sounds good, addressed in newest changes.